src/lib/customPropTypes.js

```diff
@@ -0,0 +1,182 @@
+ * @param {String[]} allowedTypes Collection of allowed component types.
+ */
+export const ofComponentTypes = (allowedTypes) => {
+  return (props, propName, componentName) => {
+    if (propName !== 'children') {
+      throw new Error(`ofComponentTypes can only be used on the \`children\` prop, not ${propName}.`)
+    }
+    if (!_.isArray(allowedTypes)) {
+      throw new Error([
+        'Invalid argument supplied to ofComponentTypes, expected an instance of array.'
+          ` See ${componentName} prop \`${propName}\`.`,
+      ].join(''))
+    }
+    const disallowed = _.compact(Children.map(props.children, child => {
+      return _.includes(allowedTypes, _.get(child, 'type._meta.name')) ? null : child
+    }))
+    if (!_.isEmpty(disallowed)) {
+      return new Error(
+        `\`${componentName}\` should only have children of type \`${allowedTypes}\`.`
+      )
+    }
+  }
+}
+
+/**
+ * Verifies exclusivity of a given prop.
+ * @param {string[]} disallowedProps An array of props that cannot be used with this prop.
+ */
+export const disallow = disallowedProps => {
+  return (props, propName, componentName) => {
+    if (!_.isArray(disallowedProps)) {
+      throw new Error([
+        'Invalid argument supplied to mutuallyExclusive, expected an instance of array.'
+          ` See ${componentName} prop \`${propName}\`.`,
+      ].join(''))
+    }
+
+    // mutually exclusive
+    const disallowed = disallowedProps.reduce((acc, exclusive) => {
+      if (!_.isUndefined(props[propName]) && !_.isUndefined(props[exclusive])) {
+        return [...acc, exclusive]
+      }
+      return acc
+    }, [])
+
+    if (!_.isEmpty(disallowed)) {
+      return new Error([
+        `\`${componentName}\` prop \`${propName}\` conflicts with props: \`${disallowed.join('`, `')}\`.`,
+        'They cannot be defined together, choose one or the other.',
+      ].join(' '))
+    }
+  }
+}
+
+/**
+ * Ensure a prop adherers to multiple prop type validators.
+ * @param {function[]} validators An array of propType functions.
+ */
+export const every = (validators) => {
+  return (props, propName, componentName, ...rest) => {
+    if (!_.isArray(validators)) {
+      throw new Error([
+        'Invalid argument supplied to all, expected an instance of array.',
+        `See ${componentName} prop \`${propName}\`.`,
+      ].join(' '))
+    }
+
+    const errors = _.compact(_.map(validators, validator => {
+      if (!_.isFunction(validator)) {
+        throw new Error(
+          `all() argument "validators" should contain functions, found: ${type(validator)}.`
+        )
+      }
+      return validator(props, propName, componentName, ...rest)
+    }))
+
+    // we can only return one error at a time
+    return errors[0]
+  }
+}
+
+/**
+ * Ensure a prop adherers to at least one of the given prop type validators.
+ * @param {function[]} validators An array of propType functions.
+ */
+export const some = (validators) => {
```

Comment:
Renamed `any` => `some` as `PropTypes.any` already exists and means something different.  `some` here has parity with Array#some.